test/shopify-cli/task/ensure_env_test.rb

```diff
@@ -14,18 +14,148 @@ def setup
-
-      def test_writes_env_if_no_env
-        CLI::UI.expects(:ask).times(3)
-          .returns('apikey', 'apisecret', 'https://test-shop.myshopify.com')
-        @task = EnsureEnv.call(@context)
-        assert_equal('apikey', Project.current.env.api_key)
+      def test_create_new_app_if_none_available
+        response = [{
+          "id" => 421,
+          "businessName" => "one",
+          "stores" => [{
+            "shopDomain" => "store.myshopify.com",
+          }],
+          "apps" => [],
+        }]
+        ShopifyCli::PartnersAPI::Organizations.stubs(:fetch_with_app).returns(response)
+        CLI::UI::Prompt.expects(:ask).with(@context.message('core.tasks.ensure_env.app_name')).returns('new app')
+        CLI::UI::Prompt.expects(:ask).with(@context.message('core.tasks.ensure_env.app_type.select')).returns('public')
+        ShopifyCli::Tasks::CreateApiClient.expects(:call).with(
+          @context,
+          org_id: 421,
+          title: 'new app',
+          type: 'public',
+        ).returns({
+          "apiKey" => "ljdlkajfaljf",
+          "apiSecretKeys" => [{ "secret" => "kldjakljjkj" }],
+          "id" => "12345678",
+        })
+        Resources::EnvFile.any_instance.stubs(:write)
+        assert_equal response.first, EnsureEnv.call(@context)
+      end
+
+      def test_no_prompt_if_one_app_and_org
+        response = [{
+          "id" => 421,
+          "businessName" => "one",
+          "stores" => [{
+            "shopDomain" => "store.myshopify.com",
+          }],
+          "apps" => [{
+            "title" => "app",
+            "apiKey" => 1234,
+            "apiSecretKeys" => [{
+              "secret" => 1233,
+            }],
+          }],
+        }]
+        ShopifyCli::PartnersAPI::Organizations.stubs(:fetch_with_app).returns(response)
+        Resources::EnvFile.any_instance.stubs(:write)
+        assert_equal response.first, EnsureEnv.call(@context)
+      end
+
+      def test_uses_default_values_for_env_file
+        Resources::EnvFile.expects(:parse_external_env).with.raises(Errno::ENOENT)
+        ShopifyCli::PartnersAPI::Organizations.stubs(:fetch_with_app).returns(partners_api_response)
+        expect_user_prompts
+        Resources::EnvFile.expects(:new).with(env_file_values).returns(stub(:write))
+        assert_equal partners_api_response[1], EnsureEnv.call(@context)
+      end
+
+      def test_keep_existing_env_values
+        Resources::EnvFile.expects(:parse_external_env).with.returns({ host: "host" })
+        ShopifyCli::PartnersAPI::Organizations.stubs(:fetch_with_app).returns(partners_api_response)
+        expect_user_prompts
+        Resources::EnvFile.expects(:new).with(env_file_values.merge({ host: "host" })).returns(stub(:write))
+        assert_equal partners_api_response[1], EnsureEnv.call(@context)
+      end
+
+      def test_not_all_required_values_found
+        Resources::EnvFile.expects(:parse_external_env).with.returns({ api_key: "apikey", secret: "secret" })
```

Comment:
It's so much cleaner with `env_file_values`.
nitpick: I realized that `env_file_values` doesn't need to be a function, and can be a variable as well.  Up to you.  We can probably move `{api_key:..., secret:...}` to a similar function/variable.  Call it `existing_env_file_value`, the other one `new_env_file_value`...or something?